unstable test TestResourceGroupRunaway

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
 [2023/06/21 18:38:34.445 +08:00] [INFO] [db.go:625] ["Waiting for closer"]
--- FAIL: TestResourceGroupRunaway (31.28s)
    resource_group_test.go:255: 
        	Error Trace:	ddl/tests/resourcegroup/resource_group_test.go:255
        	Error:      	"[[rg1 select /*+ resource_group(rg1) */ * from t 2023-06-21 18:38:31]]" should have 0 item(s), but has 1
        	Test:       	TestResourceGroupRunaway
FAIL
```
### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

--- a/ddl/tests/resourcegroup/resource_group_test.go
+++ b/ddl/tests/resourcegroup/resource_group_test.go
@@ -245,38 +245,32 @@ func TestResourceGroupRunaway(t *testing.T) {
 	err := tk.QueryToErr("select /*+ resource_group(rg1) */ * from t")
 	require.ErrorContains(t, err, "[executor:8253]Query execution was interrupted, identified as runaway query")
 
-	// consifer the low speed of write in test, all check will exec after sleeping 1s.
-	delayDuration := time.Millisecond * 750
-	time.Sleep(time.Millisecond*10 + delayDuration)
-	tk.MustQuery("select SQL_NO_CACHE resource_group_name, original_sql, match_type from mysql.tidb_runaway_queries").
-		Check(testkit.Rows("rg1 select /*+ resource_group(rg1) */ * from t identify"))
-	// wait for GC, because of FROM_UNIXTIME is second level, so wait for 1s.
-	time.Sleep(time.Millisecond*1000 + delayDuration)
-	require.Len(t, tk.MustQuery("select SQL_NO_CACHE resource_group_name, original_sql, time from mysql.tidb_runaway_queries").Rows(), 0)
-
+	tryInterval := time.Millisecond * 200
+	maxWaitDuration := time.Second * 5
+	tk.EventuallyMustQueryAndCheck("select SQL_NO_CACHE resource_group_name, original_sql, match_type from mysql.tidb_runaway_queries", nil,
+		testkit.Rows("rg1 select /*+ resource_group(rg1) */ * from t identify"), maxWaitDuration, tryInterval)
+	// require.Len(t, tk.MustQuery("select SQL_NO_CACHE resource_group_name, original_sql, time from mysql.tidb_runaway_queries").Rows(), 0)
+	tk.EventuallyMustQueryAndCheck("select SQL_NO_CACHE resource_group_name, original_sql, time from mysql.tidb_runaway_queries", nil,
+		nil, maxWaitDuration, tryInterval)
 	tk.MustExec("alter resource group rg1 RU_PER_SEC=1000 QUERY_LIMIT=(EXEC_ELAPSED='100ms' ACTION=COOLDOWN)")
-	tk.MustQuery("select /*+ resource_group(rg1) */ * from t")
+	tk.MustQuery("select /*+ resource_group(rg1) */ * from t").Check(testkit.Rows("1"))
 
 	tk.MustExec("alter resource group rg1 RU_PER_SEC=1000 QUERY_LIMIT=(EXEC_ELAPSED='100ms' ACTION=DRYRUN)")
-	tk.MustQuery("select /*+ resource_group(rg1) */ * from t")
+	tk.MustQuery("select /*+ resource_group(rg1) */ * from t").Check(testkit.Rows("1"))
 
 	err = tk.QueryToErr("select /*+ resource_group(rg2) */ * from t")
 	require.ErrorContains(t, err, "Query execution was interrupted, identified as runaway query")
 	tk.MustGetErrCode("select /*+ resource_group(rg2) */ * from t", mysql.ErrResourceGroupQueryRunawayQuarantine)
-	time.Sleep(time.Millisecond*10 + delayDuration)
-	tk.MustQuery("select SQL_NO_CACHE resource_group_name, original_sql, match_type from mysql.tidb_runaway_queries").
-		Check(testkit.Rows(
-			"rg2 select /*+ resource_group(rg2) */ * from t identify",
-			"rg2 select /*+ resource_group(rg2) */ * from t watch",
-		))
+	tk.EventuallyMustQueryAndCheck("select SQL_NO_CACHE resource_group_name, original_sql, match_type from mysql.tidb_runaway_queries", nil,
+		testkit.Rows("rg2 select /*+ resource_group(rg2) */ * from t identify",
+			"rg2 select /*+ resource_group(rg2) */ * from t watch"), maxWaitDuration, tryInterval)
 	tk.MustQuery("select SQL_NO_CACHE resource_group_name, watch_text from mysql.tidb_runaway_quarantined_watch").
 		Check(testkit.Rows("rg2 select /*+ resource_group(rg2) */ * from t"))
 
-	time.Sleep(time.Millisecond*1000 + delayDuration)
-	require.Len(t, tk.MustQuery("select SQL_NO_CACHE resource_group_name, original_sql, match_type from mysql.tidb_runaway_queries").Rows(), 0)
-	// watch duration is 1s
-	time.Sleep(time.Millisecond*1000 + delayDuration)
-	require.Len(t, tk.MustQuery("select SQL_NO_CACHE resource_group_name, watch_text, end_time from mysql.tidb_runaway_quarantined_watch").Rows(), 0)
+	tk.EventuallyMustQueryAndCheck("select SQL_NO_CACHE resource_group_name, original_sql, time from mysql.tidb_runaway_queries", nil,
+		nil, maxWaitDuration, tryInterval)
+	tk.EventuallyMustQueryAndCheck("select SQL_NO_CACHE resource_group_name, watch_text, end_time from mysql.tidb_runaway_quarantined_watch", nil,
+		nil, maxWaitDuration, tryInterval)
 
 	tk.MustExec("alter resource group rg2 RU_PER_SEC=1000 QUERY_LIMIT=(EXEC_ELAPSED='50ms' ACTION=COOLDOWN)")
 	tk.MustQuery("select /*+ resource_group(rg2) */ * from t").Check(testkit.Rows("1"))

--- a/testkit/result.go
+++ b/testkit/result.go
@@ -49,6 +49,21 @@ func (res *Result) Check(expected [][]interface{}) {
 	res.require.Equal(needBuff.String(), resBuff.String(), res.comment)
 }
 
+// Equal check whether the result equals the expected results.
+func (res *Result) Equal(expected [][]interface{}) bool {
+	resBuff := bytes.NewBufferString("")
+	for _, row := range res.rows {
+		_, _ = fmt.Fprintf(resBuff, "%s\n", row)
+	}
+
+	needBuff := bytes.NewBufferString("")
+	for _, row := range expected {
+		_, _ = fmt.Fprintf(needBuff, "%s\n", row)
+	}
+
+	return bytes.Equal(needBuff.Bytes(), resBuff.Bytes())
+}
+
 // AddComment adds the extra comment for the Result's output.
 func (res *Result) AddComment(c string) {
 	res.comment += "\n" + c

--- a/testkit/testkit.go
+++ b/testkit/testkit.go
@@ -149,6 +149,22 @@ func (tk *TestKit) MustQuery(sql string, args ...interface{}) *Result {
 	return tk.MustQueryWithContext(context.Background(), sql, args...)
 }
 
+// EventuallyMustQueryAndCheck query the statements and assert that
+// result rows.lt will equal the expected results in waitFor time, periodically checking equality each tick.
+// Note: retry can't ignore error of the statements. If statements returns error, it will break out.
+func (tk *TestKit) EventuallyMustQueryAndCheck(sql string, args []interface{},
+	expected [][]interface{}, waitFor time.Duration, tick time.Duration) {
+	defer func() {
+		if tk.alloc != nil {
+			tk.alloc.Reset()
+		}
+	}()
+	tk.require.Eventually(func() bool {
+		res := tk.MustQueryWithContext(context.Background(), sql, args...)
+		return res.Equal(expected)
+	}, waitFor, tick)
+}
+
 // MustQueryWithContext query the statements and returns result rows.
 func (tk *TestKit) MustQueryWithContext(ctx context.Context, sql string, args ...interface{}) *Result {
 	comment := fmt.Sprintf("sql:%s, args:%v", sql, args)

